src/onegov/activity/iso20022.py

```diff
@@ -128,6 +128,45 @@ def transaction_entries(root: 'etree._Element') -> 'Iterator[etree._Element]':
         yield entry
 
 
+def get_esr(booking_text: str) -> str | None:
+
+    """
+    Extracts the QR-bill reference number from the given text.
+    QR-bill reference numbers can be 26 or 27 characters long.
+    The 26-character version doesn't include the check digit.
+
+    For example:
+
+    input: 'Gutschrift QRR: 27 99029 05678 18860 27295 37059'
+    output: '269902905678188602729537059'
+
+    :returns: The extracted reference number or None if no reference number.
+    If the extracted reference number is only 26 characters long, the check
+    digit is appended to the end.
+    """
+
+    # Pattern for 26-digit ESR numbers
+    pattern_26 = r'\b(\d{2}\s*?\d{5}\s*?\d{5}\s*?\d{5}\s*?\d{5}\s*?\d{4})\b'
+
+    # Pattern for 27-digit ESR numbers
+    pattern_27 = r'\b(\d{2}\s*?\d{5}\s*?\d{5}\s*?\d{5}\s*?\d{5}\s*?\d{5})\b'
+
+    # Try matching 27-digit pattern first
+    match = re.search(pattern_27, booking_text)
+    if match:
+        return re.sub(r'\s', '', match.group(1))
+
+    # If no 27-digit match, try 26-digit pattern
+    match = re.search(pattern_26, booking_text)
+    if match:
+        ref26 = re.sub(r'\s', '', match.group(1))
+        # append checksum as it is stored like this in the database
+        return ref26 + ESRSchema().checksum(ref26)
```

Comment:
If the booking text contains a number with a length == 26, it seems to me that there are two possibilities. 
1. They don't send the checksum.
2. They strip one leading zero.

Currently we only account for (1), but not (2). Is (2) impossible and I'm missing something?
If not, this is actually non-trivial.

Reply:
Right, shorter reference numbers are possible (leading zeros), https://github.com/OneGov/onegov-cloud/blob/master/src/onegov/activity/models/invoice_reference.py#L284
But then the `get_esr()` would also fail when extracting it....